In [4]:
import numpy as np

# Ejercicio 1: Operaciones Matriciales
Dada una matriz en formato numpy array, donde cada fila de la matriz representa un vector matemático:

Computar las normas l0, l1, l2, l-infinito 

In [55]:
A = np.random.randint(0, 10, (5, 4))
print(A)

[[1 9 1 3]
 [7 8 4 1]
 [1 4 6 9]
 [3 0 3 4]
 [5 4 9 0]]


#### Norma 0 (I0)

In [56]:
def norma_0(matriz):
    """Calcula la norma cero de los vectores filas de la matriz
    que recibe. Es decir, devuelve la cantidad de valores distintos
    de cero que hay en cada vector"""
    return np.count_nonzero(matriz, axis=1)

#### Norma p
<p style='text-align: justify;'>
    Devuelve la noma:
</p>

<p style='text-align: center;'>
    $||x||_p = \Large (\sum \limits _{i=1} ^{n} |x_i|^p)^{1/p}$
</p>

In [20]:
#def norma_p(matriz, p=1):
    #"""Calcula la norma p de los vectores
    #filas de la matriz que recibe."""
    #return np.power(np.sum(np.power(np.abs(matriz),p),axis=1),1/p)

def norma_p(matriz, p=1, ax = 1):
    """Calcula la norma p de los vectores
    filas de la matriz que recibe."""
    return np.power(np.sum(np.power(np.abs(matriz),p),axis=ax),1/p)

#### Norma infinita $\infty$ (I-infinito)
<p style='text-align: center;'>
    $||x||_\infty =  max_i |x_i|$
</p>

In [58]:
def normainf(matriz):
    """Calcula la norma infinito de los vectores filas de la matriz
    que recibe. Es decir, devuelve el elemento de mayor valor que 
    hay en cada vector"""
    return np.amax(matriz,axis = 1)

#### Adicional: Deteccion de elemento de valor minimo

In [59]:
def minimo(matriz):
    return np.amin(matriz,axis = 1)

In [60]:
n_0 = norma_0(A)
n_1 = norma_p(A, 1)
n_2 = norma_p(A, 2)
n_inf = normainf(A)
for vector in range(A.shape[0]):
    print(f"Vector: {A[vector,:]}")
    print(f"Norma 0 (I0): {n_0[vector]}")
    print(f"Norma 1 (I1): {n_1[vector]}")
    print(f"Norma 2 (I2): {n_2[vector]}")
    print(f"Norma inf (I∞): {n_inf[vector]}\n")

Vector: [1 9 1 3]
Norma 0 (I0): 4
Norma 1 (I1): 14.0
Norma 2 (I2): 9.591663046625438
Norma inf (I∞): 9

Vector: [7 8 4 1]
Norma 0 (I0): 4
Norma 1 (I1): 20.0
Norma 2 (I2): 11.40175425099138
Norma inf (I∞): 8

Vector: [1 4 6 9]
Norma 0 (I0): 4
Norma 1 (I1): 20.0
Norma 2 (I2): 11.575836902790225
Norma inf (I∞): 9

Vector: [3 0 3 4]
Norma 0 (I0): 3
Norma 1 (I1): 10.0
Norma 2 (I2): 5.830951894845301
Norma inf (I∞): 4

Vector: [5 4 9 0]
Norma 0 (I0): 3
Norma 1 (I1): 18.0
Norma 2 (I2): 11.045361017187261
Norma inf (I∞): 9



# Ejercicio 2: Sorting
<p style='text-align: justify;'>
    Dada una matriz en formato numpy array, donde cada fila de la matriz representa un vector matemático, se requiere computar la norma l2 de cada vector. Una vez obtenida la norma, se debe ordenar las mísmas de mayor a menor. Finalmente, obtener la matriz original ordenada por fila según la norma l2.


Todas las operaciones debe ser vectorizadas.
</p>

In [61]:
def sort_of_vectors(matriz, p, show = False):
    """Ordena de mayor a menor segun el
    valor de la norma p de los  vectores
    filas de la matriz que recibe
    * Devuelve la matriz con sus filas ordenadas segun la norma-2,
    * El arreglo de indices de la matriz original ordenado segun la norma-2,
    * El arreglo de norma-2 de la matriz original"""
    n2 = norma_p(matriz, p)       #Genero un vector con las normas-2 de cada vector
    orden = np.argsort(n2)[::-1]  #Genero un vector con los indices de manera ordenada segun la norma-2
    matriz_sort = matriz[orden]   #Genero la matriz ordenada segun la norma-2 de los vectores filas
    if show == True:
        # Muestro el orden original
        print("Orden original")
        for vector in range(matriz.shape[0]):
            print(f"Vector: {matriz[vector,:]} \t Norma-{p}: {n2[vector]}")
        print("\nOrdenado segun su norma 2")
        for vector in range(matriz.shape[0]):
            print(f"Vector: {matriz_sort[vector,:]} \t Norma-{p}: {n2[orden[vector]]}")
    return matriz_sort, orden, n2

In [62]:
# Declaro la matriz
B = np.random.randint(0, 10, (5, 4))
print(B)

[[2 0 4 2]
 [1 5 0 5]
 [8 2 3 7]
 [8 2 2 6]
 [2 1 9 0]]


In [63]:
p = 2
BS, orden, n2 = sort_of_vectors(B, p, show = True)

Orden original
Vector: [2 0 4 2] 	 Norma-2: 4.898979485566356
Vector: [1 5 0 5] 	 Norma-2: 7.14142842854285
Vector: [8 2 3 7] 	 Norma-2: 11.224972160321824
Vector: [8 2 2 6] 	 Norma-2: 10.392304845413264
Vector: [2 1 9 0] 	 Norma-2: 9.273618495495704

Ordenado segun su norma 2
Vector: [8 2 3 7] 	 Norma-2: 11.224972160321824
Vector: [8 2 2 6] 	 Norma-2: 10.392304845413264
Vector: [2 1 9 0] 	 Norma-2: 9.273618495495704
Vector: [1 5 0 5] 	 Norma-2: 7.14142842854285
Vector: [2 0 4 2] 	 Norma-2: 4.898979485566356


# Ejercicio 3: Indexing
<p style='text-align: justify;'>
    El objetivo es construir un índice para identificadores de usuarios, es decir id2idx e idx2id. Para ello crear una clase, donde el índice se genere en el constructor. Armar métodos get_users_id y get_users_idx.

    - Identificadores de usuarios : users_id  = [15, 12, 14, 10, 1, 2, 1]
    - Índice de usuarios :          users_idx = [ 0,  1,  2,  3, 4, 5, 4]
    
    id2idx =  [-1     4     5    -1    -1    -1     -1    -1    -1    -1     3     -1      1    -1     2     0]
              [ 0     1     2     3     4     5      6     7     8     9    10     11     12    13    14    15]

    id2idx[15] -> 0 ; id2idx[12] -> 1 ; id2idx[3] -> -1
    idx2id[0] -> 15 ; idx2id[4] -> 1
    
</p>

In [64]:
class UsersIndex:
    def __init__(self,users_ids):
        ids, idxs = np.unique(users_ids, return_index=True)
        idx2 = np.full((np.max(ids)+1),-1)
        idx2[ids]=idxs
        
        self.idx2id = idx2
        self.id2idx = np.arange(ids.max()+1)
        print(f"ID:\t{self.id2idx}")
        print(f"Indice:\t{self.idx2id}")
        
        
    def get_users_id(self, index):
        if index != -1:
            print(f"idx2id[{index}]->{self.id2idx[np.argwhere(self.idx2id == index)[0,0]]}")
        else:
            print(f"idx2id[{index}]->{self.id2idx[np.argwhere(self.idx2id == index)[0,0]]}")

    def get_users_idx(self, identification):
        print(f"id2idx[{identification}]->{self.idx2id[np.argwhere(self.id2idx == identification)[0,0]]}")
        
adm = UsersIndex([15, 12, 14, 10, 1, 2, 1])
print()
adm.get_users_id(0)
adm.get_users_id(4)
print()
adm.get_users_idx(15)
adm.get_users_idx(12)
adm.get_users_idx(3)
adm.get_users_idx(1)

ID:	[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]
Indice:	[-1  4  5 -1 -1 -1 -1 -1 -1 -1  3 -1  1 -1  2  0]

idx2id[0]->15
idx2id[4]->1

id2idx[15]->0
id2idx[12]->1
id2idx[3]->-1
id2idx[1]->4


# Ejercicio 4: Precision, Recall, Accuracy
<p style='text-align: justify;'>
    En los problemas de clasificación, se cuenta con dos arreglos, la verdad (ground truth) y la predicción (prediction). Cada elemento de los arreglos puede tomar dos valores: True (representado por 1) y False (representado por 0). Por lo tanto, se pueden definir cuatro variables:

- True Positive (TP): la verdad es 1 y la predicción es 1.
- True Negative (TN): la verdad es 0 y la predicción es 0.
- False Negative (FN): la verdad es 1 y la predicción es 0.
- False Positive (FP): la verdad es 0 y la predicción es 1.

A partir de esas cuatro variables, se definen las siguientes métricas:

- Precision = TP / (TP + FP)
- Recall = TP / (TP + FN)
- Accuracy = (TP + TN) / (TP + TN + FP + FN)

Para los siguientes arreglos, representando la verdad y la predicción, calcular las métricas anteriores con operaciones vectorizadas en NumPy.

* truth = [1,1,0,1,1,1,0,0,0,1]
* prediction = [1,1,1,1,0,0,1,1,0,0]
</p>

In [65]:
truth = [1, 1, 0, 1, 1, 1, 0, 0, 0, 1]
pred = [1, 1, 1, 1, 0, 0, 1, 1, 0, 0]
print(truth)
print(pred)

[1, 1, 0, 1, 1, 1, 0, 0, 0, 1]
[1, 1, 1, 1, 0, 0, 1, 1, 0, 0]


In [66]:
def PRA(truth, pred):
    # Calculo el valor True Positive
    TP = sum(np.logical_and(truth,pred))
    # Calculo el valor True Negative
    TN = sum(np.logical_and(np.logical_not(truth),np.logical_not(pred)))
    # Calculo el valor False Negative
    FN = sum(np.logical_and(truth,np.logical_not(pred)))
    # Calculo el valor False Negative
    FP = sum(np.logical_and(np.logical_not(truth),pred))
    PRECISION = TP / (TP + FP)
    RECALL = TP / (TP + FN)
    ACCURACY = (TP + TN) / (TP + TN + FP + FN)
    
    return TP, TN, FN, FP, PRECISION, RECALL, ACCURACY

In [67]:
TP, TN, FN, FP, PRECISION, RECALL, ACCURACY = PRA(truth, pred)
print(f"Cantidad de Verdaderos Positivos: {TP}")
print(f"Cantidad de Verdaderos Negativos: {TN}")
print(f"Cantidad de Falsos Negativos: {FN}")
print(f"Cantidad de Falsos Positivos: {FP}")
print(f"Precision: {PRECISION}")
print(f"Precision: {RECALL}") #sensibilidad?
print(f"Precision: {ACCURACY}")

Cantidad de Verdaderos Positivos: 3
Cantidad de Verdaderos Negativos: 1
Cantidad de Falsos Negativos: 3
Cantidad de Falsos Positivos: 3
Precision: 0.5
Precision: 0.5
Precision: 0.4


# Ejecicio 5: Average Query Precision
<p style='text-align: justify;'>
    En information retrieval o search engines, en general contamos con queries “q” y para cada “q” una lista de documentos que son verdaderamente relevantes. Para evaluar un search engine, es común utilizar la métrica average query precision. Tomando de referencia el siguiente ejemplo, calcular la métrica con NumPy utilizando operaciones vectorizadas.

    q_id            = [1, 1, 1, 1, 2, 2, 2, 3, 3, 3, 3, 3, 4, 4, 4, 4]
    predicted_rank  = [0, 1, 2, 3, 0, 1, 2, 0, 1, 2, 3, 4, 0, 1, 2, 3]
    truth_relevance = [T, F, T, F, T, T, T, F, F, F, F, F, T, F, F, T] 

* Precision para q_id 1 = 2 / 4
* Precision para q_id 2 = 3 / 3
* Precision para q_id 3 = 0 / 5
* Precision para q_id 4 = 2 / 4

Average query precision = ((2/4) + (3/3) + (0/5) + (2/4)) / 4
</p>

In [5]:
def queryprecision(q_id, truth_relevance):
    ids = np.unique(q_id)
    
    q_id_expanded = np.zeros((q_id.size, ids.size))
    q_id_expanded = q_id_expanded + q_id[:, np.newaxis]
    
    truth_relevance_expanded = np.zeros((truth_relevance.size, ids.size))
    truth_relevance_expanded = truth_relevance_expanded + truth_relevance[:, np.newaxis]
    
    relevance = np.sum((q_id_expanded == ids) & (truth_relevance_expanded == True),axis=0)
    solicitudes = np.sum(q_id_expanded==ids,axis=0)
    
    return np.sum(relevance / solicitudes)/ids.size

In [6]:
q_id            = np.array([1, 1, 1, 1, 2, 2, 2, 3, 3, 3, 3, 3, 4, 4, 4, 4])
predicted_rank  = np.array([0, 1, 2, 3, 0, 1, 2, 0, 1, 2, 3, 4, 0, 1, 2, 3])
truth_relevance = np.array([True, False, True, False, True, True, True, 
                            False,False, False, False, False, True, False, 
                            False, True])

queryprecision(q_id, truth_relevance)

0.5

# Ejecicio 6: Distancia a centroides
<p style='text-align: justify;'>
    Dada una nube de puntos X y centroides C, obtener la distancia entre cada vector X y los centroides utilizando operaciones vectorizadas y broadcasting en NumPy. Utilizar como referencia los siguientes valores:
    
    X = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    C = [[1, 0, 0], [0, 1, 1]]
   
</p>

In [56]:
x=np.array([[1,2,3],[4,5,6],[7,8,9]])
c=np.array([[1,0,0],[0,1,1]])

c1=c[:,np.newaxis]

distances = norma_p(c1-x, p=2, ax = 2)
print()

for centroide in range(c.shape[0]):
    for punto in range(x.shape[0]):
        print(f"Distancia entre {x[punto, :]} y {c[centroide, :]} es {distances[centroide, punto]}. Diferencia {(c1-x)[centroide, punto]}")


Distancia entre [1 2 3] y [1 0 0] es 3.605551275463989. Diferencia [ 0 -2 -3]
Distancia entre [4 5 6] y [1 0 0] es 8.366600265340756. Diferencia [-3 -5 -6]
Distancia entre [7 8 9] y [1 0 0] es 13.45362404707371. Diferencia [-6 -8 -9]
Distancia entre [1 2 3] y [0 1 1] es 2.449489742783178. Diferencia [-1 -1 -2]
Distancia entre [4 5 6] y [0 1 1] es 7.54983443527075. Diferencia [-4 -4 -5]
Distancia entre [7 8 9] y [0 1 1] es 12.727922061357855. Diferencia [-7 -7 -8]


# Ejecicio 7: Distancia a centroides - parte 2
<p style='text-align: justify;'>
    Obtener para cada fila en X, el índice de la fila en C con distancia euclídea más pequeña.
    Es decir, decir para cada fila en X a qué cluster pertenece en C.
    Por ejemplo, si el resultado anterior fue:

        # [[ 3.60555128 8.36660027 13.45362405]
        # [ 2.44948974 7.54983444 12.72792206]]
    
    El programa debería devolver [1, 1, 1]
   
</p>

In [55]:
min_distances = np.argmin(distances,axis=0)
print()
print(min_distances)


[1 1 1]


# Ejecicio 8: Implementacion basica de K-means en Numpy
<p style='text-align: justify;'>
   K-means es uno de los algoritmos más básicos en Machine Learning no supervisado. Es un algoritmo de clusterización, que agrupa los datos que comparten características similares. Recordemos que entendemos datos como n realizaciones del vector aleatorio X.
    
   El algoritmo K-means funciona de la siguiente manera:
    
    1- El usuario selecciona la cantidad de clusters a crear (n).
    2- Se seleccionan n elementos aleatorios de X como posiciones iniciales del los centroides C.
    3- Se calcula la distancia entre todos los puntos en X y todos los puntos en C.
    4- Para cada punto en X se selecciona el centroide más cercano de C.
    5- Se recalculan los centroides C a partir de usar las filas de X que pertenecen a cada centroide. 
    6- Se itera entre 3 y 5 una cantidad fija de veces o hasta que la posición de los centroides no cambie.
    
   Implementar la función def k_means(X, n) de manera tal que al finalizar devuelva la posición de los centroides y a qué cluster pertenece cada fila de X.

   Hint: para (2) utilizar funciones de np.random, para (3) y (4) usar los ejercicios anteriores, para (5) es válido utilizar un for. Iterar 10 veces entre (3) y (5).
   
</p>

# Ejecicio 9: Computar todas las métricas con __call__
<p style='text-align: justify;'>
   En problemas de machine learning, es muy común que para cada predicción que obtenemos en nuestro dataset de verificación y evaluación, almacenemos en arreglos de numpy array el resultado de dicha predicción, junto con el valor verdadero y parámetros auxiliares (como el ranking de la predicción y el query id).
    
   Luego de obtener todas las predicciones, podemos utilizar la información almacenada en los arreglos de numpy, para calcular todas las métricas que queremos medir en nuestro sistema. 
    
   Una buena práctica para implementar esto en Python, es crear clases que hereden de una clase Metric “base” y que cada métrica implemente el método __call__.
   
   Utilizar herencia, operador __call__, kwargs, para escribir un programa que permita calcular todas las métricas de los ejercicios anteriores mediante un for. Cuál es la ventaja de resolver el problema utilizando éstas herramientas?
   
</p>

# Ejecicio 10: Transformar un dataset a numpy estructurado con singleton
<p style='text-align: justify;'>
   Descargar un dataset de ejemplo de la siguiente URL: https://www.kaggle.com/rounakbanik/the-movies-dataset/data?select=ratings.csv
   Crear la cuenta en Kaggle, porque es un lugar de donde potencialmente vamos a descargar muchos recursos.
   Crear un array estructurado en numpy que represente correctamente la estructura del dataset (“userId”, “movieId”, “rating”, “timestamp”).
   Crear una clase que permita:
    
* Crear la estructura de un structured numpy array para el dataset.
* Leer el csv, almacenar la información en el array estructurado.
* Guardar el array estructurado en formato .pkl
* Crear una instancia singleton del array estructurado (utilizando __new__ e __init__).
* Al crear la instancia, si se encuentra el .pkl cargar desde el pkl. Si el .pkl no está, comenzar por transformar el .csv en .pkl y luego levantar la información.
* Encontrar una forma de optimizar la operación usando generators.
   
</p>